pkg/knit/cmd/ethtool/ethtool.go

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ */
+
+package ethtool
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/spf13/cobra"
+
+	"github.com/openshift-kni/debug-tools/pkg/knit/cmd"
+	goethtool "github.com/safchain/ethtool"
+)
+
+type ethtoolOptions struct {
+	showFeatures bool
+	showChannels bool
+}
+
+func NewEthtoolCommand(knitOpts *cmd.KnitOptions) *cobra.Command {
+	opts := &ethtoolOptions{}
+	eInfo := &cobra.Command{
+		Use:   "ethtool",
+		Short: "subset of ethtool query capabilities",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return showEthtool(cmd, opts, args)
+		},
+		Args: cobra.MaximumNArgs(1),
+	}
+	eInfo.Flags().BoolVarP(&opts.showFeatures, "show-features", "k", false, "show the features of the selected devices")
+	eInfo.Flags().BoolVarP(&opts.showChannels, "show-channels", "l", false, "show the channels of the selected devices")
+	return eInfo
+}
+
+func showEthtool(cmd *cobra.Command, opts *ethtoolOptions, args []string) error {
+	var err error
+	ifaces := args
+	if len(ifaces) == 0 {
+		ifaces, err = getAllInterfaceNames()
+	}
+	if err != nil {
+		return err
+	}
+
+	ethHandle, err := goethtool.NewEthtool()
+	if err != nil {
+		return err
+	}
+	defer ethHandle.Close()
+
+	for _, iface := range ifaces {
+		if opts.showFeatures {
+			feats, err := ethHandle.Features(iface)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("Features for %s:\n", iface)
+			for key, val := range feats {
+				fmt.Printf("%s: %s\n", key, toggle(val))
+			}
+		}
+
+		if opts.showChannels {
+			chans, err := ethHandle.GetChannels(iface)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("Channel parameters for %s:\n", iface)
+			fmt.Printf("Pre-set maximums:\n")
+			fmt.Printf("RX:		%d\n", chans.MaxRx)
+			fmt.Printf("TX:		%d\n", chans.MaxTx)
+			fmt.Printf("Other:		%d\n", chans.MaxOther)
+			fmt.Printf("Combined:	%d\n", chans.MaxCombined)
+			fmt.Printf("Current hardware settings:\n")
+			fmt.Printf("RX:		%d\n", chans.RxCount)
+			fmt.Printf("TX:		%d\n", chans.TxCount)
+			fmt.Printf("Other:		%d\n", chans.OtherCount)
+			fmt.Printf("Combined:	%d\n", chans.CombinedCount)
+			fmt.Printf("\n")
+		}
+	}
+	return nil
+}
+
+func toggle(v bool) string {
+	if v {
+		return "on"
+	}
+	return "off"
+}
+
+func getAllInterfaceNames() ([]string, error) {
+	ifaces, err := net.Interfaces()
```

Comment:
This will add all the interfaces right ?  I wonder if there is a need to split between supported devices and non supported .  
I would also test what happens when non supported devices are in the mix and if we can get an ethtool error that for one of them resulting in command exit.

Reply:
the problem is indeed distinguish between supported and unsupported devices. In case operation is not supported (= listing channels) the program will indeed abort when hitting the specific interface.